test/jdk/java/foreign/capturecallstate/TestCaptureCallState.java

```diff
@@ -34,99 +34,181 @@
 
 import java.lang.foreign.Arena;
 import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.GroupLayout;
 import java.lang.foreign.Linker;
 import java.lang.foreign.MemoryLayout;
 import java.lang.foreign.MemorySegment;
 import java.lang.foreign.StructLayout;
+import java.lang.foreign.ValueLayout;
 import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
 import java.lang.invoke.VarHandle;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.function.Consumer;
 
 import static java.lang.foreign.MemoryLayout.PathElement.groupElement;
+import static java.lang.foreign.ValueLayout.ADDRESS;
 import static java.lang.foreign.ValueLayout.JAVA_DOUBLE;
 import static java.lang.foreign.ValueLayout.JAVA_INT;
 import static java.lang.foreign.ValueLayout.JAVA_LONG;
 import static org.testng.Assert.assertEquals;
 
 public class TestCaptureCallState extends NativeTestHelper {
 
+    static final MethodHandle MH_UPCALL_TARGET;
+
     static {
         System.loadLibrary("CaptureCallState");
-        if (IS_WINDOWS) {
-            String system32 = System.getenv("SystemRoot") + "\\system32";
-            System.load(system32 + "\\Kernel32.dll");
-            System.load(system32 + "\\Ws2_32.dll");
+
+        try {
+            MH_UPCALL_TARGET = MethodHandles.lookup().findStatic(TestCaptureCallState.class, "upcallTarget",
+                    MethodType.methodType(Object.class, MemorySegment.class, VarHandle.class, int.class, Object.class));
+        } catch (ReflectiveOperationException e) {
+            throw new ExceptionInInitializerError(e);
         }
     }
 
-    private record SaveValuesCase(String nativeTarget, FunctionDescriptor nativeDesc, String threadLocalName, Consumer<Object> resultCheck) {}
+    private record SaveValuesCase(String nativeTarget, String threadLocalName, Optional<MemoryLayout> retValueLayout) {
+        SaveValuesCase(String nativeTarget, String threadLocalName) {
+            this(nativeTarget, threadLocalName, Optional.empty());
+        }
+        SaveValuesCase(String nativeTarget, String threadLocalName, MemoryLayout retValueLayout) {
+            this(nativeTarget, threadLocalName, Optional.of(retValueLayout));
+        }
+    }
 
     @Test(dataProvider = "cases")
-    public void testSavedThreadLocal(SaveValuesCase testCase) throws Throwable {
+    public void testDowncalls(SaveValuesCase testCase) throws Throwable {
         Linker.Option stl = Linker.Option.captureCallState(testCase.threadLocalName());
-        MethodHandle handle = downcallHandle(testCase.nativeTarget(), testCase.nativeDesc(), stl);
+        FunctionDescriptor downcallDesc = testCase.retValueLayout()
+                .map(rl -> FunctionDescriptor.of(rl, JAVA_INT, rl))
+                .orElse(FunctionDescriptor.ofVoid(JAVA_INT));
```

Comment:
Using [`Optional::orElseGet(…)`] avoids allocating the `FunctionDescriptor` corresponding to `(int)void` when `testCase.retValueLayout()` is present:
```suggestion
                .orElseGet(() -> FunctionDescriptor.ofVoid(JAVA_INT));
```

[`Optional::orElseGet(…)`]: https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/Optional.html#orElseGet(java.util.function.Supplier)